src/setup_runs/config_read_functions.py

```diff
@@ -0,0 +1,311 @@
+
+    Parameters
+    ----------
+    config
+        The configuration dictionary.
+
+    Returns
+    -------
+        The updated configuration dictionary after variable substitutions.
+        The number of iterations performed for substitutions.
+    """
+    avail_keys = list(config.keys())
+    iterationCount = 0
+    while iterationCount < 10 :
+        ## check if any entries in the config dictionary need populating
+        foundToken = False
+        for key, value in config.items() :
+            if isinstance(value, str) :
+                if (value.find('${') >= 0) :
+                    foundToken = True
+        ##
+        if foundToken :
+            for avail_key in avail_keys :
+                key = '${%s}' % avail_key
+                value = config[avail_key]
+                for k in avail_keys :
+                    if isinstance(config[k], str) :
+                        if config[k].find(key) >= 0 :
+                            config[k] = config[k].replace(key, value)
+        else :
+            break
+        ##
+        iterationCount += 1
+
+    return config, iterationCount
+
+
+def parse_boolean_keys(config: dict,
+                       truevals: list[str] = ['true', '1', 't', 'y', 'yes'],
+                       falsevals: list[str] = ['false', '0', 'f', 'n', 'no'],
+                       bool_keys: list[str] = ["run_as_one_job", "submit_wrf_now", "submit_wps_component",
+                                               "only_edit_namelists",
+                                               "restart",
+                                               'delete_metem_files', "use_high_res_sst_data",
+                                               'regional_subset_of_grib_data']
+                       ) :
+    boolvals = truevals + falsevals
+
+    for bool_key in bool_keys :
+        assert (
+                config[bool_key].lower() in boolvals
+        ), f'Key {bool_key} not a recognised boolean value'
+        config[bool_key] = (config[bool_key].lower() in truevals)
+
+    return config
+
+
+## function to parse times
+def process_date_string(datestring) :
+    datestring = datestring.strip().rstrip()
+
+    ## get the timezone
+    if len(datestring) <= 19 :
+        tz = pytz.UTC
+        date = datetime.datetime.strptime(datestring, '%Y-%m-%d %H:%M:%S')
+    else :
+        tzstr = datestring[20 :]
+        tz = pytz.timezone(tzstr)
+        date = datetime.datetime.strptime(datestring, '%Y-%m-%d %H:%M:%S %Z')
+
+    date = tz.localize(date)
+
+    return date
+
+
+def load_wrf_config(filename: str) -> WRFConfig :
+    input_str = read_config_file(filename)
+
+    config = parse_config(input_str)
+
+    config = add_environment_variables(config=config, environmental_variables=os.environ)
+
+    config, iterationCount = substitute_variables(config)
+
+    # Iteration count for filling variables
+    assert iterationCount < 10, "Config key substitution exceeded iteration limit..."
```

Comment:
This can live inside `substitute_variables` then `iterationCount` doesn't need to be returned

Reply:
Agree. Am I missing something or is it actually impossible to exceed the `iterationCount`of 10?

The logic for the iterations in `substitute_variables` is behind a `while` loop:

`while iterationCount < 10 :`

Reply:
I just triggered this `assert` statement, so it is possible. Now I can also see what it's good for. It can happen when the `HOME` directory is not created from the environment variables. Perhaps an error message that guides in this direction would be helpful.